logger/log.go

```diff
@@ -92,44 +112,128 @@ func timeEncoder(t time.Time, enc zapcore.PrimitiveArrayEncoder) {
 }
 
 func Info(args ...interface{}) {
-	log.Info(args)
+	globalLog.Info(args)
 }
 
 func Infof(format string, args ...interface{}) {
-	log.Infof(format, args...)
+	globalLog.Infof(format, args...)
 }
 
 func Debug(args ...interface{}) {
 	if !DebugEnabled {
 		return
 	}
-	log.Debug(args)
+	globalLog.Debug(args)
 }
 
 func Debugf(format string, args ...interface{}) {
-	log.Debugf(format, args...)
+	globalLog.Debugf(format, args...)
 }
 
 func Warn(args ...interface{}) {
-	log.Warn(args)
+	globalLog.Warn(args)
 }
 
 func Warnf(format string, args ...interface{}) {
-	log.Warnf(format, args...)
+	globalLog.Warnf(format, args...)
 }
 
 func Error(args ...interface{}) {
-	log.Error(args)
+	globalLog.Error(args)
 }
 
 func Errorf(format string, args ...interface{}) {
-	log.Errorf(format, args...)
+	globalLog.Errorf(format, args...)
 }
 
 func Fatal(args ...interface{}) {
-	log.Fatal(args)
+	globalLog.Fatal(args)
 }
 
 func Fatalf(format string, args ...interface{}) {
-	log.Fatalf(format, args...)
+	globalLog.Fatalf(format, args...)
+}
+
+// GetLogger gets a logger with the given name. The name is logged as a prefix for any message coming from this logger.
+func GetLogger(loggerName string) (*TektiteLogger, error) {
+	return GetLoggerWithLevel(loggerName, globalLevel)
+}
+
+func GetLoggerWithLevel(loggerName string, level zapcore.Level) (*TektiteLogger, error) {
+	initLock.Lock()
+	defer initLock.Unlock()
+	if !initialised {
+		return nil, errors.Errorf("cannot create a logger with name %s before the global logger is initialised", loggerName)
+	}
+
+	if value, exists := loggersByName[loggerName]; exists {
+		Warnf("tried to create a duplicate logger with name %s", loggerName)
+		return value, nil
+	}
+
+	logger := CreateLogger(level, globalEncoding)
+	tektiteLogger := NewTektiteLogger(logger, loggerName)
+	loggersByName[loggerName] = tektiteLogger
+	return tektiteLogger, nil
+}
+
+type TektiteLogger struct {
+	logger       *zap.Logger
+	log          *zap.SugaredLogger
+	debugEnabled bool
+	prefix       string
+}
+
+func NewTektiteLogger(logger *zap.Logger, loggerName string) *TektiteLogger {
+	log := logger.Sugar()
+
+	return &TektiteLogger{
+		logger:       logger,
+		log:          log,
+		debugEnabled: log.Desugar().Core().Enabled(zap.DebugLevel),
+		prefix:       fmt.Sprintf("[%s] ", loggerName),
+	}
+}
+
+func (t *TektiteLogger) Info(args ...interface{}) {
+	t.log.Info(append([]interface{}{t.prefix}, args...)...)
```

Comment:
I was wondering between two implementations of this - pre-allocating the slice or this one. Again through some microbenchmarks I found that this one seems better for these amounts of vars
```
// BenchmarkPreallocated-8   	 8268920	       168.0 ns/op
// BenchmarkPreallocated-8   	 7235402	       159.6 ns/op
// BenchmarkPreallocated-8   	 7508109	       155.0 ns/op

// BenchmarkAppend-8   	 7857439	       146.8 ns/op
// BenchmarkAppend-8   	 5972097	       174.6 ns/op
// BenchmarkAppend-8   	 7970244	       149.9 ns/op
func BenchmarkAppend(b *testing.B) {
	args := make([]interface{}, 100)
	name := "test"
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_ = append([]interface{}{name}, args...)
	}
}
func BenchmarkPreallocated(b *testing.B) {
	args := make([]interface{}, 100)
	name := "test"
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		newArgs := make([]interface{}, len(args)+1)
		newArgs[0] = name
		copy(newArgs[1:], args)
	}
}
```